src/Payum/Paypal/ProHosted/Nvp/Api.php

```diff
@@ -0,0 +1,270 @@
+
+    /**
+     * A payment was reversed due to a chargeback or other type of reversal.
+     * The funds have been removed from your account balance and returned to the buyer.
+     * The reason for the reversal is specified in the ReasonCode element.
+     */
+    const PAYMENTSTATUS_REVERSED = 'Reversed';
+
+    /**
+     * You refunded the payment.
+     */
+    const PAYMENTSTATUS_REFUNDED = 'Refunded';
+
+    /**
+     *  A payment has been accepted.
+     */
+    const PAYMENTSTATUS_PROCESSED = 'Processed';
+
+    const PAYERSTATUS_VERIFIED = 'verified';
+    const PAYERSTATUS_UNVERIFIED = 'unverified';
+
+    const PENDINGREASON_AUTHORIZATION = 'authorization';
+
+    const PAYMENTACTION_SALE = 'sale';
+    const FORM_CMD = '_hosted-payment';
+
+    /**
+     * @var HttpClientInterface
+     */
+    protected $client;
+
+    /**
+     * @var MessageFactory
+     */
+    protected $messageFactory;
+
+    /**
+     * @var array
+     */
+    protected $options = array(
+        'username'  => null,
+        'password'  => null,
+        'signature' => null,
+        'business'  => null,
+        'return'    => null,
+        'subtotal'  => null,
+        'sandbox'   => null,
+        'cmd'       => Api::FORM_CMD,
+    );
+
+    /**
+     * @param array               $options
+     * @param HttpClientInterface $client
+     * @param MessageFactory      $messageFactory
+     *
+     * @throws \Payum\Core\Exception\InvalidArgumentException if an option is invalid
+     */
+    public function __construct(array $options, HttpClientInterface $client, MessageFactory $messageFactory)
+    {
+        $options = ArrayObject::ensureArrayObject($options);
+        $options->defaults($this->options);
+        $options->validateNotEmpty([
+            'username',
+            'password',
+            'signature',
+            'business',
+        ]);
+
+        if (false == is_bool($options['sandbox'])) {
+            throw new InvalidArgumentException('The boolean sandbox option must be set.');
+        }
+
+        $this->options        = $options;
+        $this->client         = $client;
+        $this->messageFactory = $messageFactory;
+    }
+
+    /**
+     * Solution BMCreateButton
+     *
+     * @param array $fields
+     *
+     * @throws RuntimeException
+     *
+     * @return array
+     */
+    public function doCreateButton(array $fields)
+    {
+        if (false == isset($fields['return'])) {
+            if (false == $this->options['return']) {
+                throw new RuntimeException('The return must be set either to FormRequest or to options.');
+            }
+
+            $fields['return'] = $this->options['return'];
+        }
+
+        $fields['paymentaction'] = self::PAYMENTACTION_SALE;
+        $fields['cmd']           = self::FORM_CMD;
+
+        $newFields = [];
+        $i         = 0;
+        foreach ($fields as $key => $val) {
+            $newFields['L_BUTTONVAR'.$i] = $key.'='.$val;
+            $i++;
+        }
```

Comment:
I dont mind, I am just thinking that this is not responsibility of API to do such things. It is a logic of an action.